### PR TITLE
Added sort_index for widgets

### PR DESCRIPTION
**Pull request by saravanpa-aot**
*Issue #:*
https://github.com/bcgov/met-public/issues/<Put the github issue number here>

*Description of changes:*

- Added sort_index column
- modified logic to increase sort_index while creating a new widget
- Added new endpoint for sorting the widgets

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the met-public license (Apache 2.0).

### CODE REVIEW

File: met-api/src/met_api/services/widget_service.py [hunk 1]
```diff
@@ -32,9 +34,48 @@ def create_widget(widget_data, engagement_id, user_id):
...
+        # find the highest sort order of the engagement
+        sort_index = 0
+        widgets = WidgetModel.get_widgets_by_engagement_id(engagement_id)
+        if widgets:
+            # Find the largest in the existing widgest
+            sort_index = max(widget.sort_index for widget in widgets)
+        return sort_index
+
+    @staticmethod
+    def sort_widget(engagement_id, widget_items: list, user_id=None):
```
Comment (jadmsaadaot) [inline met-api/src/met_api/services/widget_service.py:55]: I don't think this is still widget_items it's widgets. right?

File: met-api/tests/unit/api/test_widget.py [hunk 2]
```diff
@@ -34,6 +36,81 @@ def test_create_widget(client, jwt, session, widget_info):  # pylint:disable=unu
                      headers=headers, content_type='application/json')
     assert rv.status_code == 200
 
+    rv = client.get('/api/widgets/engagement/' + str(engagement.id),
+                    headers=headers, content_type='application/json')
+    assert rv.status_code == 200
+    assert rv.json.get('result')[0].get('sort_index') == 1
+
+
+def test_create_widget_sort(client, jwt, session):  # pylint:disable=unused-argument
```
Comment (jadmsaadaot) [inline met-api/tests/unit/api/test_widget.py:45]: Too many scenarios in one test? Can they be separated? Failures here is harder to track to what scenario failed. One  scenario for each test preferred

**Review by jadmsaadaot — CHANGES_REQUESTED** [review]
Variable naming issue in widget service. And unit test design issue

File: met-api/src/met_api/services/widget_service.py (hunk 1 above)
Comment (saravanpa-aot) [reply met-api/src/met_api/services/widget_service.py:55]: its widget , we are sorting widget

File: met-api/tests/unit/api/test_widget.py (hunk 2 above)
Comment (saravanpa-aot) [reply met-api/tests/unit/api/test_widget.py:45]: I can separate the invalid test to another one.. the other test just asserts the sorting order

File: met-api/migrations/versions/e37d79be3a05_added_sort_for_widgets.py [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+# revision identifiers, used by Alembic.
+revision = 'e37d79be3a05'
+down_revision = '9f86fdcfb248'
+branch_labels = None
+depends_on = None
+
+
+def upgrade():
+    # ### Add Sort index to widget.###
+    op.add_column('widget', sa.Column('sort_index', sa.Integer(), nullable=False, server_default='1'))
```
Comment (jadmsaadaot) [inline met-api/migrations/versions/e37d79be3a05_added_sort_for_widgets.py:21]: learning from this that I don't have to do nullable True, execute then do nullable false. Thanks

File: met-api/src/met_api/services/widget_service.py (hunk 1 above)
Comment (jadmsaadaot) [reply met-api/src/met_api/services/widget_service.py:55]: so we need to change the paramater name from widget_items to widgets

**Review by tiagograf-AOT — APPROVED** [review]

File: met-api/src/met_api/services/widget_service.py (hunk 1 above)
Comment (saravanpa-aot) [reply met-api/src/met_api/services/widget_service.py:55]: yeah..i misunderstood..now its fixed

**Review by djnunez-aot — APPROVED** [review]

**Review by VineetBala-AOT — APPROVED** [review]

File: met-api/src/met_api/services/widget_service.py [hunk 4]
```diff
@@ -32,9 +34,48 @@ def create_widget(widget_data, engagement_id, user_id):
...
+            'id': widget_item.get('id'),
+            'sort_index': widget_item.get('sort_index'),
+            'updated_by': user_id
+        } for widget_item in widgets
+        ]
+
+        WidgetModel.update_widgets(widget_sort_mappings)
+
+    @staticmethod
+    def _validate_widget_ids(engagement_id, widget_items):
```
Comment (jadmsaadaot) [inline met-api/src/met_api/services/widget_service.py:69]: still calling them widget_items here

File: met-api/tests/unit/api/test_widget.py (hunk 2 above)
Comment (jadmsaadaot) [reply met-api/tests/unit/api/test_widget.py:45]: We are posting widget_info_1 and widge_info_2 here. We don't need to do that right?

If we are testing the sorting, then the sorting endpoint should be the only one called. The test has multiple API calls, it is a bit difficult to understand what it is testing. Ideally, One test is testing only one API call. Everything else should be given.

File: met-api/src/met_api/services/widget_service.py (hunk 4 above)
Comment (saravanpa-aot) [reply met-api/src/met_api/services/widget_service.py]: Fixed now

**Review by jadmsaadaot — APPROVED** [review]

File: met-api/tests/unit/api/test_widget.py (hunk 2 above)
Comment (saravanpa-aot) [reply met-api/tests/unit/api/test_widget.py:45]: Will work on this later..I will create a ticket for tests in model layer and separate the tests

**Pull request merged**